### PR TITLE
*: replace use of context.WithTimeout with a wrapper

### PR DESCRIPTION
**Pull request by jordanlewis**
context.WithTimeout is a tricky beast. If you're not careful, you'll add
timeouts to lots of contexts but keep the naive error propagation paths
we're all used to: passing up an error if we get one.

This oftentimes leads to opaque and confusing error messages for the
user and for us as developers. "context deadline exceeded" is often
unactionable, and often makes it tough to track down what actually went
wrong underneath the hood.

This patch adds a wrapper, contextutil.RunWithTimeout, that runs an
anonymous function with a timeout and returns a prettier error if the
function does in fact time out. The error will include an operation name
and the deadline that was exceeded.

This patch also removes all usages of context.WithTimeout and
context.WithDeadline with contextutil.RunWithTimeout and
contextutil.RunWithDeadline.

Also, add a lint to prevent them from creeping back in.

### CODE REVIEW

File: pkg/util/contextutil/context.go [hunk 1]
```diff
@@ -39,3 +41,62 @@ func wrap(ctx context.Context, cancel context.CancelFunc) (context.Context, cont
...
+	deadline  time.Time
+}
+
+func (t TimeoutError) Error() string {
+	if !t.deadline.IsZero() {
+		return fmt.Sprintf("operation %q missed deadline of %s", t.operation, t.deadline)
+	}
+	return fmt.Sprintf("operation %q timed out after %s", t.operation, t.duration)
+}
+func (TimeoutError) Timeout() bool   { return true }
```
Comment (mjibson) [inline pkg/util/contextutil/context.go]: this and temporary need documentation about why they exist

File: pkg/util/contextutil/context.go [hunk 2]
```diff
@@ -39,3 +41,62 @@ func wrap(ctx context.Context, cancel context.CancelFunc) (context.Context, cont
...
+func (TimeoutError) Temporary() bool { return true }
+func (TimeoutError) Cause() error {
+	// This ensures that people looking for DeadlineExceeded in particular still
+	// see it.
+	return context.DeadlineExceeded
+}
+
+// RunWithTimeout runs a function with a timeout, the same way you'd do with
+// context.WithTimeout. It improve the opaque error messages returned by
```
Comment (mjibson) [inline pkg/util/contextutil/context.go]: "improves". ditto below.

**Comment by awoods187** [comment]
I think this is awesome work!

**Review by tbg — APPROVED** [review]
:lgtm:

I'm not particularly fond of the closure pattern used here but I don't know that Go lets you do anything better. Looking forward to never seeing an undecorated `context.DeadlineExceeded` any more. Thanks for tackling this!

> Reviewed 1 of 1 files at r1, 18 of 18 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34888#-:-LYgNEM16mfrhVZ0ukd5:b8rqd8u)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, and @mjibson)

___
*[pkg/internal/client/txn.go, line 615 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34888#-LYgMG0K4M3c0wiP6YlW:-LYgMG0L5aI5xNC2nl7s:b-kjpgyo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b702f79b654d0850381bf7b2287a99246684c2bf/pkg/internal/client/txn.go#L615)):*
> ```Go
> 	// long-lived network partition).
> 	stopper := txn.db.ctx.Stopper
> 	ctx, cancel1 := stopper.WithCancelOnQuiesce(txn.db.AnnotateCtx(context.Background()))
> ```

take it or leave it, but you could rename cancel1 back to cancel.
___
*[pkg/server/status.go, line 1043 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34888#-LYgMd7d-jsqjNoI8bHB:-LYgMd7d-jsqjNoI8bHC:b-radq7x) ([raw file](https://github.com/cockroachdb/cockroach/blob/88517b84c98b716a76b2648af2f11228412850a3/pkg/server/status.go#L1043)):*
> ```Go
> 	}
> 
> 	// Subtract base.NetworkTimeout from the deadline so we have time to process
> ```

This functionality is lost, right? Seems OK. I wonder how this got used.

**Review by petermattis — COMMENTED** [review]
:lgtm: though I didn't verify all of the changes with a fine-toothed comb. 

I'm actually pleasantly surprised that the closure pattern makes some of the code more readable (IMO). The alternative to the closure pattern would be a function to wrap the error if it is a `DeadlineExceeded`:

```
func AnnotateDeadlineExceeded(ctx context.Context, err error, msg string) error {
  if err == context.DeadlineExceeded || ctx.Err() == context.DeadlineExceeded { ... }
}
```

We'd have to make sure to call this function everywhere `WithTimeout` and `WithDeadline` is used, which would make linting harder. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34888#-:-LYgSIa49YPe_gVcNzr-:b-qgo5et)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, @mjibson, and @tbg)

___
*[pkg/server/status.go, line 1043 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34888#-LYgMd7d-jsqjNoI8bHB:-LYgRmOaEJoZjub86x2M:bi0od3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/88517b84c98b716a76b2648af2f11228412850a3/pkg/server/status.go#L1043)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This functionality is lost, right? Seems OK. I wonder how this got used.
</blockquote></details>

I recall Matt T added this. I'm not so sure about removing it. @jordanlewis Can you track down the reason it was added in the commit history?

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34888)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @dt, @mjibson, and @tbg)

___
*[pkg/internal/client/txn.go, line 615 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34888#-LYgMG0K4M3c0wiP6YlW:-LYgsGz-CssR2SvJ518G:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b702f79b654d0850381bf7b2287a99246684c2bf/pkg/internal/client/txn.go#L615)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

take it or leave it, but you could rename cancel1 back to cancel.
</blockquote></details>

Done.
___
*[pkg/server/status.go, line 1043 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34888#-LYgMd7d-jsqjNoI8bHB:-LYgsKNb9J_5WoH1Xjh3:bjomsz) ([raw file](https://github.com/cockroachdb/cockroach/blob/88517b84c98b716a76b2648af2f11228412850a3/pkg/server/status.go#L1043)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I recall Matt T added this. I'm not so sure about removing it. @jordanlewis Can you track down the reason it was added in the commit history?
</blockquote></details>

The code was added in 2016 by Tristan Rice in #8546.

However, it makes no sense to me - it's *subtracting* a timeout from the deadline, if there already exists a deadline on the context, which makes the timeout closer to now. What is the point of that? It seemed safe to delete. I don't think it really did much - and in any case, the worst effect that this change will have is make the timeout 3 seconds longer.

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34888)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @dt, @mjibson, and @tbg)

___
*[pkg/server/status.go, line 1043 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34888#-LYgMd7d-jsqjNoI8bHB:-LYgvvLO7WTxnY1p2bLe:b5mw1wv) ([raw file](https://github.com/cockroachdb/cockroach/blob/88517b84c98b716a76b2648af2f11228412850a3/pkg/server/status.go#L1043)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

The code was added in 2016 by Tristan Rice in #8546.

However, it makes no sense to me - it's *subtracting* a timeout from the deadline, if there already exists a deadline on the context, which makes the timeout closer to now. What is the point of that? It seemed safe to delete. I don't think it really did much - and in any case, the worst effect that this change will have is make the timeout 3 seconds longer.
</blockquote></details>

Ack. Thanks for looking into this. Carry on.

**Review by jordanlewis — COMMENTED** [review]
So, there's one major gotcha with this - you can't use this wrapper for functions that need to escape into different goroutines, since the wrapper automatically cancels the context at the end...

Any bright ideas? I guess the escaping goroutines should have separate contexts?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34888#-:-LYhX3-EEZ9Lt9Qkd15G:but1lp1)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @dt, @mjibson, and @tbg)

___
*[pkg/util/contextutil/context.go, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34888#-LYdXemRZmudO7KDcPAI-r1-57:-LYhTbjoB34BOsRnbHtO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/986c9c82406bebefe87c99e1e86e81ff03ce0254/pkg/util/contextutil/context.go#L57)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

this and temporary need documentation about why they exist
</blockquote></details>

Done.
___
*[pkg/util/contextutil/context.go, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34888#-LYdXemRZmudO7KDcPAI-r1-66:-LYhTc4XEQ-fK2X_e5EU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/986c9c82406bebefe87c99e1e86e81ff03ce0254/pkg/util/contextutil/context.go#L66)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

"improves". ditto below.
</blockquote></details>

Done.

**Comment by jordanlewis** [comment]
Thanks for the reviews, going to merge this now.

bors r+

**Comment by andreimatei** [comment]
Hey, is this utility a good idea? The caller of the `RunWithTimeout` function will get a nicer error message, but whomever observes the timeout still gets an opaque error (`ContextDeadlineExceeded`). And if anyone wraps that error, `RunWithTimeout` is defeated. Conversely, the utility might also wrap too much: if a sub-operation's deadline fires (and that sub-op is not using this utility), then `RunWithTimeout` assigns the wrong op name to the error.

I don't think I like this utility. I think we should write our own Context impl that's not crap.

**Comment by tbg** [comment]
> I think we should write our own Context impl that's not crap.

You can't impl your own context, unfortunately. Or even worse, you can, but it won't work because the `context` package only knows its internal impls and will screw up with anything else. Or maybe you're suggesting that we write our own version of a `Context` that's not actually a `context.Context`? Either way, this change should make things better overall already, I'd say.

**Comment by jordanlewis** [comment]
@andreimatei, how would you say it's defeated? The error has exactly the same semantics as before and can be unwrapped with `errors.Cause`, just as it always could. Anyway, I would say that accepting nothing but a full rewrite of `context.Context` is too perfectionist, considering that we've known about this problem for at least a year without taking any action.

**Comment by andreimatei** [comment]
> how would you say it's defeated?

I mean if `RunWithTimeout` receives a wrapped error, it has a choice between overwriting that error (as I see it does) and not recognizing the timeout and doing nothing. Neither of these options seems satisfying.

Plus the other problem I listed - only layers above the `RunWithTimeout` would get a nicer error. The lower level operations that are actually interrupted by the timeout don't know what hit them.

So I don't really see this patch helping much.

> Or even worse, you can, but it won't work because the `context` package only knows its internal impls and will screw up with anything else.

You're thinking of the cancelation chaining? That does work with other `Context` impls, but not as efficiently. I was thinking we'd write our own `contextutil.WithCancel()`.

**Comment by jordanlewis** [comment]
It doesn't overwrite the error - it annotates it. The error behaves to callers exactly as it did before, as long as the callers are equipped to use `errors.Cause`, which everybody should, or look explicitly for the `net.Error` interface methods, which we really should be doing for stuff like this.

What do you mean "lower level operations"? This patch isn't designed to help any "lower level operations". It's designed to prevent worthless "context deadline exceeded" errors from being returned via the pgwire protocol, which happens with regularity.

**Comment by andreimatei** [comment]
> It doesn't overwrite the error - it annotates it.

Does it? If the second disjunction is taken here, the error is overwritten, isn't it?
https://github.com/cockroachdb/cockroach/blob/eaad50f0ea64985a0d7da05abb00cc9f321c5fa9/pkg/util/contextutil/context.go

**Comment by jordanlewis** [comment]
You're right - that's not great. I should have added a cause field with the original error. I can do that.

**Pull request merged**